### PR TITLE
Makefile,session: collect the benchmark data and generate json output file

### LINKED ISSUE #25754: Micro benchmark regularly and display benchmark result
## Feature Request

**Is your feature request related to a problem? Please describe:**

We do not use the benchmark well.

Benchmarks can guide us the evaluation of the product performance.

- We just cover the typical TPCH  TPCC Sysbench test which is too high level.
- We just run the benchmark tests before a new release.

But they are too high level.

**Describe the feature you'd like:**

I'd like to have some micro benchmarks and run them regularly.

For example, if we have a `make bench` command to run the micro benchmarks, and the result is collected, 
and visualized, then we know the performance regression early.

**Describe alternatives you've considered:**

For example, there is a scheme benchmark website, https://ecraven.github.io/r7rs-benchmarks/
From which we can see the performance between different scheme compile implementation easily.

Another example is the rust pref https://perf.rust-lang.org/

I'd like TiDB to have something similar, there are may micro benchmarks and the result can be compared by day.

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #25754

See the [daily benchmark output visualized](http://www.zenlife.tk:18081/)

Problem Summary:

### What is changed and how it works?

Collect the benchmark result from `session/bench_test.go`, there are many micro benchmark tests:

```
		BenchmarkBasic,
		BenchmarkTableScan,
		BenchmarkTableLookup,
		BenchmarkExplainTableLookup,
		BenchmarkStringIndexScan,
		BenchmarkExplainStringIndexScan,
		BenchmarkStringIndexLookup,
		BenchmarkIntegerIndexScan,
		BenchmarkIntegerIndexLookup,
		BenchmarkDecimalIndexScan,
		BenchmarkDecimalIndexLookup,
		BenchmarkInsertWithIndex,
		BenchmarkInsertNoIndex,
		BenchmarkSort,
		BenchmarkJoin,
		BenchmarkJoinLimit,
		BenchmarkPartitionPruning,
		BenchmarkRangeColumnPartitionPruning,
		BenchmarkHashPartitionPruningPointSelect,
		BenchmarkHashPartitionPruningMultiSelect,
```

We do not use it well.

What's Changed:

- provide a `make bench-daily TO=xxx.json` command
- add a `TestBenchDaily` function to convert the benchmark result into json

How it Works:

I choose json because the original output disrupted by the unrelated logs, see https://github.com/pingcap/tidb/pull/25796

On the TiDB side, we provide a way to generate the json output file.

The QA team is responsible to run the benchmark daily and processing the json file, store them and render the result  https://perf.pingcap.com

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

### Release note 

- No release note

### CODE REVIEW

File: session/bench_test.go [hunk 1]
```diff
@@ -1584,3 +1589,107 @@ func BenchmarkHashPartitionPruningMultiSelect(b *testing.B) {
...
+		r2 := benchmarkResultToJSON(name, r1)
+		res = append(res, r2)
+	}
+
+	if *outfile == "" {
+		*outfile = fmt.Sprintf("%s_%s.json", *date, *commitHash)
+	}
+	out, err := os.Create(*outfile)
+	if err != nil {
+		panic(err)
```
Comment (xhebox) [inline session/bench_test.go:1681]: `t.Fatal` maybe.

File: session/bench_test.go [hunk 2]
```diff
@@ -1584,3 +1589,107 @@ func BenchmarkHashPartitionPruningMultiSelect(b *testing.B) {
...
+
+	output := BenchOutput{
+		Date:   *date,
+		Commit: *commitHash,
+		Result: res,
+	}
+	enc := json.NewEncoder(out)
+	err = enc.Encode(output)
+	if err != nil {
+		panic(err)
```
Comment (xhebox) [inline session/bench_test.go:1693]: Ditto

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: session/bench_test.go [hunk 3]
```diff
@@ -1584,3 +1589,107 @@ func BenchmarkHashPartitionPruningMultiSelect(b *testing.B) {
...
+	idx := strings.LastIndexByte(fullName, '.')
+	if idx > 0 && idx+1 < len(fullName) {
+		return fullName[idx+1:]
+	}
+	return fullName
+}
+
+var (
+	date       = flag.String("date", "", " commit date")
+	commitHash = flag.String("commit", "0ec8f2d9f", "brief git commit hash")
```
Comment (tangenta) [inline session/bench_test.go:1626]:
```suggestion
	commitHash = flag.String("commit", "unknown", "brief git commit hash")
```

**Review by tangenta — APPROVED** [review]

**Pull request merged**